micrometer-tracing-bridges/micrometer-tracing-bridge-brave/src/main/java/io/micrometer/tracing/brave/bridge/W3CPropagation.java

```diff
@@ -213,49 +99,43 @@ public <R> TraceContext.Injector<R> injector(Setter<R, String> setter) {
+            Baggage baggage = this.braveBaggageManager.getBaggage(BraveTraceContext.fromBrave(context), TRACESTATE);
             if (baggage == null) {
                 return;
             }
             String traceState = baggage.get(BraveTraceContext.fromBrave(context));
             if (StringUtils.isNotBlank(traceState)) {
-                setter.put(carrier, TRACE_STATE, traceState);
-            }
-        }
-    }
-
-    private String padLeftWithZeros(String string, int length) {
-        if (string.length() >= length) {
-            return string;
-        }
-        else {
-            StringBuilder sb = new StringBuilder(length);
-            for (int i = string.length(); i < length; i++) {
-                sb.append('0');
+                setter.put(carrier, TRACESTATE, traceState);
             }
-
-            return sb.append(string).toString();
         }
     }
 
-    void copyTraceFlagsHexTo(char[] dest, int destOffset, TraceContext context) {
-        dest[destOffset] = '0';
-        dest[destOffset + 1] = Boolean.TRUE.equals(context.sampled()) ? '1' : '0';
-    }
-
     @Override
     public <R> TraceContext.Extractor<R> extractor(Getter<R, String> getter) {
         Objects.requireNonNull(getter, "getter");
         return carrier -> {
-            String traceParent = getter.get(carrier, TRACE_PARENT);
+            String traceParent = getter.get(carrier, TRACEPARENT);
             if (traceParent == null) {
                 return withBaggage(TraceContextOrSamplingFlags.EMPTY, carrier, getter);
             }
-            TraceContext contextFromParentHeader = extractContextFromTraceParent(traceParent);
+            TraceContext contextFromParentHeader = TraceparentFormat.get().parse(traceParent);
             if (contextFromParentHeader == null) {
                 return withBaggage(TraceContextOrSamplingFlags.EMPTY, carrier, getter);
+            } else {
+                // TODO: the previous implementation always set the shared flag to true. This is
+                // incorrect because it could be a messaging span which is never shared. Setting it
+                // to shared would cause message consumers to all join the same producer span and
+                // make a mess in UI and analysis. Also, some RPC spans are intentionally not
+                // shared. Finally, b3's "don't sample" "b3=0" is conflated with "not yet sampled"
+                // unless you propagate the b3 bits.
+                //
+                // W3C traceparent has no field to indicate shared status, which is why tracestate
+                // exists, to carry the extra data not possible to encode in traceparent, such as
+                // intentionally unsampled, debug and shared flags.
+                // https://github.com/openzipkin-contrib/brave-propagation-w3c has a correct impl.
+                contextFromParentHeader = contextFromParentHeader.toBuilder().shared(true).build();
```

Comment:
@jeqo imagine 100 consumers all sharing the same span ID?!

Reply:
if you take out this if, the tests from before fail. in other words, they prove the bug is here lol

Reply:
single-producer single-consumer spans will be ok even if poorly modeled with shared spans.
single-producer multi-consumer spans will be a train wreck as you'll have N hosts in the same span ID contributing different timestamps etc. When that bug happens, you'll need to remove this if statement and re-release. Otherwise, folks might end up just disabling W3C on their own.

Reply:
actually i was wrong this is not an uber bug. only a bug. this only affects using current span or fromSpan. shared is not a propagated field so fixing this bug should have no  bad impact. will do in  a followup